### PR TITLE
fix: Fix elevation of allowed destinations

### PR DESCRIPTION
**Pull request by alexanderwe**
Resolves https://github.com/tuist/tuist/issues/5978

### Short description 📝

This PR should fix an issue with the application of platform conditions causing frameworks to be linked to platforms for which they do not specify the required destinations. 

When target A with a specific set of destinations is set as a dependency to a target B which has a superset of the destinations of target A, without specifying platform conditions manually, it will be linked to all destinations of target B even though we already know these are not supported. 

### How to test the changes locally 🧐

> Include a set of steps for the reviewer to test the changes locally (see [the documentation](https://docs.tuist.io/contributors/testing-strategy) for reference).

### Contributor checklist ✅

- [ ] The code has been linted using run `mise run lint:fix`
- [ ] The change is tested via unit testing or acceptance testing, or both
- [ ] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Comment by alexanderwe** [comment]
Hi @waltflanagan, 
Since it's quite some time I contributed to the topic of multi platforms and I remember you were the main developer of this feature I wanted to ask you, if you are aware of this issue ? 
I tried to find a first solution which involves a logic as follows. If a target does not have explicit platform conditions set, it should at least propagate the conditions which are matching its defined set of destinations. 

I already tried to test my very rough first change here, but the target is somehow still ending up being linked to all destinations of target B even though I thought I have overridden the platform conditions accordingly. 

Do you maybe have an idea why this is not working or if my approach is even suitable in the whole context of multi platform ? 

Thanks in advance for any input :)

File: Sources/TuistGenerator/Generator/TargetGenerator.swift [hunk 1]
```diff
@@ -152,7 +152,8 @@ final class TargetGenerator: TargetGenerating {
                 let nativeTarget = nativeTargets[targetSpec.name]!
                 let nativeDependency = nativeTargets[dependency.target.name]!
                 let pbxTargetDependency = try nativeTarget.addDependency(target: nativeDependency)
-                pbxTargetDependency?.applyCondition(dependency.condition, applicableTo: targetSpec)
+                pbxTargetDependency?.applyCondition(dependency.getCondition(), applicableTo: targetSpec)
+                print(pbxTargetDependency)
```
Comment (freak4pc) [inline Sources/TuistGenerator/Generator/TargetGenerator.swift:156]:
```suggestion
```

File: Sources/TuistGraph/Graph/ConditionalGraphTarget.swift [hunk 2]
```diff
@@ -17,6 +17,15 @@ public struct GraphTargetReference: Equatable, Comparable, Hashable, CustomDebug
         self.condition = condition
     }
 
+    public func getCondition() -> PlatformCondition? {
+
+        let filters = target.destinations.map {
+            $0.platformFilter
+        }
+
+        return condition ?? .when(Set(filters))
```
Comment (freak4pc) [inline Sources/TuistGraph/Graph/ConditionalGraphTarget.swift:26]:
```suggestion
        condition ?? .when(Set(target.destinations.map(\.platformFilter))
```

**Review by freak4pc — COMMENTED** [review]
Drive-by review :P

**Comment by waltflanagan** [comment]
Alright, i had some time to think.  One challenging bit in this space is that not all dependency relationships are linking dependencies.  iOS Apps depend on WatchOS apps, for example.   I don't think we can make structural assumptions built into `ConditionalGraphTarget` . for the most part it should remain just values.  Ideally this information is either linted for or created earlier on in generation when constructing the graph.  I have a branch that's taking a look at the linting approach.

**Comment by waltflanagan** [comment]
I put my previous work up in #6251

**Comment by alexanderwe** [comment]
@waltflanagan I will then close this one in favour of #6251 since the common agreement seems to be to have this as a linting rule to let consumers fix this.

**Pull request closed without merging**